src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -647,7 +647,7 @@ private static bool InterfacesAreDistinct(NamedTypeSymbol type, ConsList<Symbol>
                 return true;
             }
 
-            if (typeArgument.IsPointerType() || typeArgument.IsRestrictedType())
+            if (typeArgument.IsPointerType() || typeArgument.IsRestrictedType() || typeArgument.SpecialType == SpecialType.System_Void)
```

Comment:
I don't know how to make this happen with the current inference rules and the new error that is given on the return statement. However, this is a language constraint that might trigger in error scenarios or due to future language changes, so I believe it needs to be implemented here.